### DISCUSSION TITLE
Hyper-V Windows Server 2019 - EGL Driver message (Critical) eglInitialize: Internal Vulkan error (-3)

### DISCUSSION THREAD

**Discussion by nirm-imd** [cefsharp/CefSharp#4359] [General]
Hi,

We are initializing CefSharp (latest v108.4.13) using the following settings:
```

            var cefSettings = new CefSettings();
            cefSettings.RegisterScheme(new CefCustomScheme
            {
                SchemeName = "xx",
                SchemeHandlerFactory = _cefCustomSchema
            });

            if (!Cef.IsInitialized)
            {
                Cef.Initialize(cefSettings);
            }
```

Sporadically the CefSharp browser control doesn't display anything even though when I invoke the "GetMainFrame()?.ViewSource()" I received the proper HTML data I expected to get.
Recreating the CefSharp control (with the same settings) and\or refreshing it doesn't help.
I also tried disabling the graphics acceleration without success.

So far I was able to see the problem on virtual machines only.

Anyone encountered something similar or have any idea on how to attempt a fix?

Thank you for your time reading this query.

**Comment by amaitland** [comment]
Additional information required, please answer the following:

- Are you using `WPF/WinForms/OffScreen`?
- What OS are you using?
- What version of .NET?
- Are there any errors in the log file? See https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#log-file

> So far I was able to see the problem on virtual machines only.

How does loading google.com or similar behave?

**Comment by nirm-imd** [comment]
Thank you very much for your reply.

As to your questions:
1. WPF
2. Windows Server 2019 Datacenter
3. 4.72
4. EGL Driver message (Critical) eglInitialize: Internal Vulkan error (-3): Initialization of an object could not be completed for implementation-specific reasons, in ..\..\third_party\angle\src\libANGLE\renderer\vulkan\RendererVk.cpp, initialize:1430

Further info: the system is a Hyper-V machine


Thanks to your reference to the debug log (which I didn't expect to even exist) and the error in it I continued my research and for the time being I'm under the impression that initializing Cef with command line argument "disable-gpu" may have solved the issue.

I would prefer to understand the implications of disabling the GPU before I implement this as a formal solution.

If anyone can shed any light on this aspect I would be very grateful.

Thank you for your time and assistance.

**Reply by amaitland** [reply]
> > 4\. EGL Driver message (Critical) eglInitialize: Internal Vulkan error (-3): Initialization of an object could not be completed for implementation-specific reasons, 

Did you add an [app.manifest](https://github.com/cefsharp/CefSharp/wiki/Quick-Start-For-MS-.Net-4.x#40-add-appmanifest-to-your-application) as per [Post Install Instructions](https://github.com/cefsharp/CefSharp/blob/cefsharp/108/NuGet/Readme.txt#L13)?

It's required for `Chromium` to correctly detect `GPU` at runtime.

> I would prefer to understand the implications of disabling the GPU before I implement this as a formal solution.

Adding an `app.manifest` with `compatibility entries` should be sufficient for `Chromium` to detect `GPU` at runtime. Best to leave enabled if possible.

**Reply by nirm-imd** [reply]
Thank you for replying again and so fast.
I did have manifest file inside the bin folder with the supported OS as described in the link you specified.

Maybe Windows 2019 Server Datacenter isn't one of the listed compatibilities?
I couldn't find a specific ID for that system.

**Reply by amaitland** [reply]
> I did have manifest file inside the bin folder with the supported OS as described in the link you specified.

As in you have a file called `app.manifest` in your `bin folder`? That suggests you made a mistake when adding the file.

By default `.Net` will embed the `app.manifest` in your `exe` as a resource. This is the recommended approach.

Under **Visual Studio** select your project:
- Add -> New Item
- Select **Application Manifest File (Windows Only)**

An `<ApplicationManifest/>` entry should have been added to your proj file .e.g.

```xml
<PropertyGroup>
    <ApplicationManifest>app.manifest</ApplicationManifest>
</PropertyGroup>
```

See also https://learn.microsoft.com/en-us/windows/win32/sbscs/application-manifests#file-location

> Maybe Windows 2019 Server Datacenter isn't one of the listed compatibilities?
> I couldn't find a specific ID for that system.

`Windows Server 2019` uses the same `GUID` as `Windows 10`.

>{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a} -> Windows 10, Windows 11, Windows Server 2016, Windows Server 2019 and Windows Server 2022

As per https://learn.microsoft.com/en-us/windows/win32/sbscs/application-manifests#supportedos

**Reply by nirm-imd** [reply]
Thank you for replying again with useful information, it's greatly appreciated.

> As in you have a file called `app.manifest` in your `bin folder`? That suggests you made a mistake when adding the file.
As in I have "[application_name].exe.manifest" file in the folder next to the "[application_name].exe" file.
Just like it describes in the link you gave me:
> See also https://learn.microsoft.com/en-us/windows/win32/sbscs/application-manifests#file-location

> As per https://learn.microsoft.com/en-us/windows/win32/sbscs/application-manifests#supportedos
Thank you so much - I was looking for something like that :)

**Reply by amaitland** [reply]
> As in I have "[application_name].exe.manifest" file in the folder next to the "[application_name].exe" file.
> Just like it describes in the link you gave me:

The recommended option is to embed the manifest, that's what I use successfully. 

https://github.com/cefsharp/CefSharp.MinimalExample can be used for testing, it's kept up to date and has the recommended configuration.

If you are using `CefSharp` in a commercial application, please ask your company to [support it's development](https://github.com/sponsors/amaitland).

**Reply by nirm-imd** [reply]
> The recommended option is to embed the manifest, that's what I use successfully.

Embedding the application manifest (instead of having it separately) has no effect.

> https://github.com/cefsharp/CefSharp.MinimalExample can be used for testing, it's kept up to date and has the recommended configuration.

The minimal example works as expected but my application doesn't - maybe it has something to do with "headless" (that's where I got the "disable-gpu' idea from which fixed the issue).
 
> If you are using `CefSharp` in a commercial application, please ask your company to [support it's development](https://github.com/sponsors/amaitland).

I forwarded your request with my endorsement - the decision is not up to me though.

**Reply by amaitland** [reply]
> Embedding the application manifest (instead of having it separately) has no effect.

Did you add an `Application Manifest` via the `Visual Studio Add -> New Item` menu? 

> The minimal example works as expected but my application doesn't - maybe it has something to do with "headless" (that's where I got the "disable-gpu' idea from which fixed the issue).

`WPF` and `OffScreen` both use the same `Offscreen Rendering(OSR) Engine`. If the `CefSharp.MinimalExample.Wpf` works as expected then leaving `GPU` acceleration enabled on your `Hyper-V Virtual Machine` is possible (which is what I expected).

I suspect the problem lies with your application manifest.

You can call [IsWindows10OrGreater](https://learn.microsoft.com/en-us/windows/win32/api/versionhelpers/nf-versionhelpers-iswindows10orgreater) which will only return `true` if you have the correct entry in your manifest. There's a `c#` example at https://stackoverflow.com/a/38534174/4583726

> I forwarded your request with my endorsement - the decision is not up to me though.

Understood, thank you.

**Reply by nirm-imd** [reply]
> Did you add an `Application Manifest` via the `Visual Studio Add -> New Item` menu?

I did - just like you described above.

> `WPF` and `OffScreen` both use the same `Offscreen Rendering(OSR) Engine`. If the `CefSharp.MinimalExample.Wpf` works as expected then leaving `GPU` acceleration enabled on your `Hyper-V Virtual Machine` is possible (which is what I expected).
> 
> I suspect the problem lies with your application manifest.
> 
> You can call [IsWindows10OrGreater](https://learn.microsoft.com/en-us/windows/win32/api/versionhelpers/nf-versionhelpers-iswindows10orgreater) which will only return `true` if you have the correct entry in your manifest. There's a `c#` example at https://stackoverflow.com/a/38534174/4583726

IsWindows10OrGreater (as the example you referenced) returns true.
